### ISSUE TITLE
train gives variable has no variation warning for factor variable with many levels - does train one hot encode factors automatically?

### ISSUE THREAD

**Issue by kransom14** [topepo/caret#1051]
I am using train to tune a parameter grid for a gbm model. I have one factor with about 100 levels, though some levels only occur a couple times. I get a warning from train that the variable has no variation, but it seems to be one-hot encoding the variable because on its own the variable has plenty of variation, but not if it were to be one-hot encoded. For example, below, I only have six predictor variables, but the warning refers to "variable 120" and so on. The documentation states no preprocessing is done to the data in train, please clarify. 

### Minimal reproducible data:
```
library(DALEX)
library(gbm)
library(caret)

# load in the training data
train <- apartments
names(apartments)

# make a factor variable with many levels, make sure some levels only have a few instances
many_levels <- c(rep(seq(1,98,by = 1), times = 10), rep(seq(101,110,by = 1), times = 2))
apartments$many_levels <- as.factor(many_levels)
levels(apartments$many_levels)
```

### Minimal reproducible code
```
# set up the tuning grid
grid <-  expand.grid(n.trees = 500, interaction.depth = c(5,10), shrinkage = c(.002,.02),n.minobsinnode=10) 

# set up train control
fitControl <- trainControl(method = "repeatedcv", number = 10, repeats = 1, returnData = TRUE, verboseIter = FALSE)  

# run gbm within caret
tune <- train(m2.price~., data = apartments,
              method = "gbm",
              bag.fraction = .5,
              distribution = "gaussian",
              metric = "RMSE",
              tuneGrid = grid, 
              trControl = fitControl,
              na.action = na.pass,
              verbose = FALSE)
```
### Warning
```
Warning messages:
1: In (function (x, y, offset = NULL, misc = NULL, distribution = "bernoulli",  :
  variable 120: many_levels110 has no variation.
2: In (function (x, y, offset = NULL, misc = NULL, distribution = "bernoulli",  :
  variable 120: many_levels110 has no variation.
3: In (function (x, y, offset = NULL, misc = NULL, distribution = "bernoulli",  :
  variable 120: many_levels110 has no variation.
4: In (function (x, y, offset = NULL, misc = NULL, distribution = "bernoulli",  :
  variable 120: many_levels110 has no variation.
5: In (function (x, y, offset = NULL, misc = NULL, distribution = "bernoulli",  :
  variable 115: many_levels105 has no variation.
6: In (function (x, y, offset = NULL, misc = NULL, distribution = "bernoulli",  :
  variable 115: many_levels105 has no variation.
7: In (function (x, y, offset = NULL, misc = NULL, distribution = "bernoulli",  :
  variable 115: many_levels105 has no variation.
8: In (function (x, y, offset = NULL, misc = NULL, distribution = "bernoulli",  :
  variable 115: many_levels105 has no variation.
```
### Session info
```
R version 3.5.3 (2019-03-11)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows >= 8 x64 (build 9200)

Matrix products: default

locale:
[1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                           LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] caret_6.0-84    ggplot2_3.1.0   lattice_0.20-38 gbm_2.1.5       DALEX_0.2.7    

loaded via a namespace (and not attached):
 [1] foreach_1.4.4      splines_3.5.3      prodlim_2018.04.18 gtools_3.8.1       shiny_1.2.0        assertthat_0.2.1   expm_0.999-3      
 [8] stats4_3.5.3       sp_1.3-1           highr_0.7          pdp_0.7.0          yaml_2.2.0         LearnBayes_2.15.1  ipred_0.9-8       
[15] pillar_1.3.1       glue_1.3.0         digest_0.6.18      promises_1.0.1     recipes_0.1.4      colorspace_1.4-1   htmltools_0.3.6   
[22] httpuv_1.4.5.1     Matrix_1.2-17      plyr_1.8.4         timeDate_3043.102  klaR_0.6-14        pkgconfig_2.0.2    breakDown_0.1.6   
[29] questionr_0.7.0    gmodels_2.18.1     purrr_0.3.1        xtable_1.8-3       mvtnorm_1.0-10     scales_1.0.0       gdata_2.18.0      
[36] gower_0.2.0        later_0.8.0        lava_1.6.5         tibble_2.0.1       proxy_0.4-23       combinat_0.0-8     generics_0.0.2    
[43] ggpubr_0.2         withr_2.1.2        nnet_7.3-12        ALEPlot_1.1        agricolae_1.3-0    lazyeval_0.2.1     survival_2.44-1.1 
[50] magrittr_1.5       crayon_1.3.4       mime_0.6           deldir_0.1-16      nlme_3.1-138       MASS_7.3-51.3      class_7.3-15      
[57] data.table_1.12.0  tools_3.5.3        stringr_1.4.0      munsell_0.5.0      cluster_2.0.8      compiler_3.5.3     e1071_1.7-0.1     
[64] rlang_0.3.1        classInt_0.3-1     units_0.6-2        grid_3.5.3         iterators_1.0.10   rstudioapi_0.9.0   miniUI_0.1.1.1    
[71] boot_1.3-20        ModelMetrics_1.2.2 codetools_0.2-16   gtable_0.2.0       DBI_1.0.0          reshape2_1.4.3     AlgDesign_1.1-7.3 
[78] R6_2.4.0           lubridate_1.7.4    yaImpute_1.0-31    gridExtra_2.3      knitr_1.22         dplyr_0.8.0.1      factorMerger_0.3.6
[85] spdep_1.0-2        stringi_1.3.1      Rcpp_1.0.0         rpart_4.1-13       sf_0.7-3           spData_0.3.0       tidyselect_0.2.5  
[92] xfun_0.5           coda_0.19-2    
```

**Comment by topepo** [comment]
Using the formula method with `train()` will produce dummy variables (although technically not one-hot encodings since all but one levels have dummy columns). Using the non-formula interface (`train(x, y)`) will leave them alone. 

If you want to keep the dummy variables (maybe check [this out](https://bookdown.org/max/FES/categorical-trees.html)), use the zero-variance preprocessor `preProc = "zv"` to eliminate them before they are given to the model. 

Also, `train()` can use recipes as inputs and they have many more options to handling categorical predictors (see [here](https://tidymodels.github.io/recipes/articles/Dummies.html)) including novel levels, etc

**Comment by kransom14** [comment]
Thanks!

If I leave them alone using the `train(x,y)` interface, how does `train()` handle a situation where some levels are present in the cross validation testing data, but not in the training data? For example if I use 10 fold cross validation, there could be cases where some levels are only present in the 10 percent hold out data for some folds.

**Comment by topepo** [comment]
So if you have the predictors as factors, it will at least know the full set of columns to make either way. 

Suppose 

 * level "a" has no data in the training set, then `preProc = "zv"` will remove it and the model never gets it. That means that when it is in the test set, the model doesn't get that information. 

 * If "a" is in the training set but not the test set, everything should work fine. 

`recipes` has `step_other()` and `step_novel()` that are more systematic approaches to solving this (as well as `step_zv()`). Also the `embed` package has some other recipe steps with [supervised solutions](https://bookdown.org/max/FES/categorical-supervised-encoding.html) to this problem.

**Comment by kransom14** [comment]
I am not sure I understand because if I use the `x,y` specification, I should not have any zero variance predictors, but `preProcess="zv"` seems to remove two:
```
traindata <- apartments[,-1]
trainresponse <- apartments[,1]

tune <- train(traindata, trainresponse,
              method = "gbm",
              bag.fraction = .5,
              distribution = "gaussian",
              metric = "RMSE",
              preProcess = "zv",
              tuneGrid = grid, 
              trControl = fitControl,
              verbose = FALSE)

 tune$preProcess
Created from 1000 samples and 2 variables

Pre-processing:
  - ignored (2)
```
There might levels in the resampled training set that are not in the testing set for some resamples, but if the variables are not converted to dummies, what is `preProcess="zv"` doing in those cases?

**Comment by topepo** [comment]
`ignored (2)` means that it did not process two columns for pre-processing. 

> what is `preProcess="zv"` doing in those cases?

Nothing, unless you have data with the same value for other reason. It looks at all predictors.

**Comment by kransom14** [comment]
Got it, preProcess = "zv" is ignoring the factor variables. Thanks for your help!

**Comment by topepo** [comment]
Yes. If you need a more general filter, use recipes and `step_zv()`.

**Comment by kransom14** [comment]
If I do the x,y specification to keep the factor variables as is: 

```
traindata <- apartments[,-1]
trainresponse <- apartments[,1]

tune <- train(traindata, trainresponse,
              method = "gbm",
              bag.fraction = .5,
              distribution = "gaussian",
              metric = "RMSE",
              tuneGrid = grid, 
              trControl = fitControl,
              verbose = FALSE)
```
with the 10 fold cross-validation and some levels of my factor variable are not in a resample (but end up in the out of bag/testing data) for a particular fold, will train throw an error? gbm will throw an error if a new level is present in the testing data. 

Is there a way to see what data was pulled for each resample?

**Comment by topepo** [comment]
I'm not sure what `gbm` will do. `rpart` and others include all categories (based on the factor levels) in the rule. I've never seen a warning or error to that effect but it is worth testing.

After the model fit, `tune$control$index` indicates which rows of the training set were used in each fold to train the model.

**Comment by kransom14** [comment]
Thanks!

I did:

```
idxs <- tune$control$index
for(i in 1:10){
print(all(unique(apartments$many_levels) %in% unique(apartments$many_levels[idxs[[i]]])))
}
```

and there were indeed some levels of my many-leveled factor that were in the 90% training set but not the 10% testing set for several folds. This doesn't seem to bother gbm because I didn't get an error. I found this: https://github.com/gbm-developers/gbm and believe gbm sends new/unseen factor levels to the NA node.
